File: Python/0221-m-maximal-square.py
"""
DP
if matrix[i][j] == 1
    f[i][j] = min(f[i - 1][j], f[i][j-1], f[i-1][j-1]) + 1;
if matrix[i][j] == 0
    f[i][j] = 0
Time: O(mn)
Space: O(mn)
"""
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        if not matrix:
            return 0
        rows = len(matrix)
        cols = len(matrix[0])
        dp = [[0 for c in range(cols + 1)] for r in range(rows+ 1)]
        maxsqlen = 0
        for i in range(1, rows + 1):
            for j in range(1, cols + 1):
                if matrix[i - 1][j - 1] == '1':
                    dp[i][j] = min(dp[i][j - 1], dp[i - 1][j], dp[i - 1][j - 1]) + 1
                    maxsqlen = max(maxsqlen, dp[i][j])
        return maxsqlen * maxsqlen

File: leetcode/C++/0221-m-maximal-square.cpp
// DP
class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        if (matrix.empty() || matrix[0].empty())
            return 0;
        int n_row = matrix.size(), n_col = matrix[0].size();
        vector<vector<int>> dp(n_row, vector<int>(n_col, 0));
        
        int res = 0;
        for (int r = 0; r < n_row; r++) {
            for (int c = 0; c < n_col; c++) {
                dp[r][c] = matrix[r][c] - '0';
                if (r > 0 && c > 0 && dp[r][c] == 1) {
                    dp[r][c] += min(min(dp[r-1][c-1], dp[r][c-1]), dp[r-1][c]);
                }
                res = max(res, dp[r][c]);
            }
        }
        return res * res;
    }
};